File: ios/Fishers/Models/Selection.swift
import Foundation

/// Where a player sits in the selection for one fixture.
enum SelectionState: String, Codable, CaseIterable {
    case pool
    case selected
    case reserve
    case notSelected = "not_selected"
    case confirmed
    case declined
    case dropped

    var label: String {
        switch self {
        case .pool: return "Available pool"
        case .selected: return "Selected"
        case .reserve: return "Reserve"
        case .notSelected: return "Not selected"
        case .confirmed: return "Confirmed"
        case .declined: return "Declined"
        case .dropped: return "Dropped"
        }
    }

    var isInSquad: Bool { self == .selected || self == .confirmed }
}

struct SelectionCandidate: Codable, Identifiable, Equatable {
    let userId: UUID
    let name: String
    let position: String?
    let skillLevel: String?
    let availability: AvailabilityStatus?
    let reliabilityScore: Int
    let reliabilityBand: String
    /// Fixtures they were available for but left out of, last 60 days.
    let gamesMissedOut: Int
    let state: SelectionState
    let isConfirmed: Bool

    var id: UUID { userId }

    enum CodingKeys: String, CodingKey {
        case name, position, availability, state
        case userId = "user_id"
        case skillLevel = "skill_level"
        case reliabilityScore = "reliability_score"
        case reliabilityBand = "reliability_band"
        case gamesMissedOut = "games_missed_out"
        case isConfirmed = "is_confirmed"
    }
}

struct RankedCandidate: Codable, Identifiable, Equatable {
    let userId: UUID
    let name: String
    let score: Int
    /// Plain-English reasons, in the order the ranking applied them.
    let reasons: [String]

    var id: UUID { userId }

    enum CodingKeys: String, CodingKey {
        case name, score, reasons
        case userId = "user_id"
    }
}

struct PositionQuota: Codable, Equatable {
    let position: String
    let minimum: Int
}

struct SquadRequirements: Codable, Equatable {
    let size: Int
    let reserves: Int
    let positionQuotas: [PositionQuota]

    enum CodingKeys: String, CodingKey {
        case size, reserves
        case positionQuotas = "position_quotas"
    }
}

/// Everything the captain's selection screen needs in one call.
struct SelectionBoard: Codable, Equatable {
    let eventId: UUID
    let title: String
    let sport: String
    let startsAt: Date
    let status: String
    let statusNote: String?
    let requirements: SquadRequirements
    /// `off` | `suggest` | `auto_publish`
    let autonomy: String
    let confirmLeadHours: Int
    let dropLeadHours: Int
    let candidates: [SelectionCandidate]
    let ranked: [RankedCandidate]
    let selectedCount: Int
    let confirmedCount: Int

    enum CodingKeys: String, CodingKey {
        case title, sport, status, requirements, autonomy, candidates, ranked
        case eventId = "event_id"
        case startsAt = "starts_at"
        case statusNote = "status_note"
        case confirmLeadHours = "confirm_lead_hours"
        case dropLeadHours = "drop_lead_hours"
        case selectedCount = "selected_count"
        case confirmedCount = "confirmed_count"
    }

    func candidates(in state: SelectionState) -> [SelectionCandidate] {
        candidates.filter { $0.state == state }
    }

    var pool: [SelectionCandidate] {
        // Ranked order, restricted to players not yet decided.
        let undecided = Set(candidates.filter { $0.state == .pool }.map(\.userId))
        return ranked.compactMap { ranked in
            candidates.first { $0.userId == ranked.userId && undecided.contains(ranked.userId) }
        }
    }

    func reasons(for userId: UUID) -> [String] {
        ranked.first { $0.userId == userId }?.reasons ?? []
    }

    var isAssistantAvailable: Bool { autonomy != "off" }
}

/// A squad awaiting publish, from either the ranking or the assistant.
struct SquadProposal: Codable, Equatable {
    /// `ranking` or `assistant`
    let source: String
    let selected: [RankedCandidate]
    let reserves: [RankedCandidate]
    let unmetQuotas: [String]
    let announcement: String?
    let concerns: String?
    let confidence: String?
    let published: Bool

    enum CodingKeys: String, CodingKey {
        case source, selected, reserves, announcement, concerns, confidence, published
        case unmetQuotas = "unmet_quotas"
    }

    var isFromAssistant: Bool { source == "assistant" }
}

struct OutstandingFee: Codable, Identifiable, Equatable {
    let userId: UUID
    let name: String
    let eventId: UUID
    let fixture: String
    let startAt: Date
    let amountCents: Int?
    let currency: String
    let remindersSent: Int

    var id: String { "\(userId)-\(eventId)" }

    enum CodingKeys: String, CodingKey {
        case name, fixture, currency
        case userId = "user_id"
        case eventId = "event_id"
        case startAt = "start_at"
        case amountCents = "amount_cents"
        case remindersSent = "reminders_sent"
    }
}

struct OutstandingFees: Codable, Equatable {
    let totalCents: Int
    let count: Int
    let owed: [OutstandingFee]

    enum CodingKeys: String, CodingKey {
        case count, owed
        case totalCents = "total_cents"
    }
}

File: web/src/lib/selection.ts
/// Picking a side (`backend/domain/src/selection.rs`).
///
/// Two signals about whether somebody can play, and they mean different
/// things: `availability` is their standing calendar — "I'm generally around on
/// Sundays" — and `rsvp` is their answer to *this* fixture. The direct answer
/// is the one a captain picks off; the calendar is a weaker, standing hint.

export type SelectionState =
  | "pool"
  | "selected"
  | "reserve"
  | "not_selected"
  | "confirmed"
  | "declined"
  | "dropped";

export type AvailabilityStatus = "available" | "maybe" | "unavailable";
export type RsvpStatus = "going" | "not_going" | "maybe" | "invited";

export type Candidate = {
  user_id: string;
  name: string;
  position: string | null;
  skill_level: string | null;
  /// Their general calendar for that date, if they keep one.
  availability: AvailabilityStatus | null;
  /// Their answer to this fixture.
  rsvp: RsvpStatus | null;
  reliability_score: number;
  reliability_band: string;
  /// Fixtures they were available for but left out of, last 60 days.
  games_missed_out: number;
  state: SelectionState;
  is_confirmed: boolean;
};

export type RankedCandidate = {
  user_id: string;
  name: string;
  score: number;
  /// Plain-English reasons, in the order they were applied.
  reasons: string[];
};

export type PositionQuota = { position: string; minimum: number };

export type SquadRequirements = {
  size: number;
  reserves: number;
  /// Advisory: quotas are filled first, then the best remaining players.
  position_quotas: PositionQuota[];
};

export type SelectionBoard = {
  event_id: string;
  title: string;
  sport: string;
  starts_at: string;
  status: string;
  status_note: string | null;
  requirements: SquadRequirements;
  /// `off` | `suggest` | `auto_publish`
  autonomy: string;
  confirm_lead_hours: number;
  drop_lead_hours: number;
  candidates: Candidate[];
  ranked: RankedCandidate[];
  selected_count: number;
  confirmed_count: number;
};

/// A squad waiting on a captain, from the ranking or from the assistant.
export type SquadProposal = {
  /// `ranking` when the deterministic model produced it, `assistant` when the
  /// model did.
  source: string;
  selected: RankedCandidate[];
  reserves: RankedCandidate[];
  unmet_quotas: string[];
  announcement: string | null;
  concerns: string | null;
  confidence: string | null;
  /// True when club policy had it published immediately.
  published: boolean;
};

export const STATE_LABEL: Record<SelectionState, string> = {
  pool: "In the pool",
  selected: "Picked",
  reserve: "Reserve",
  not_selected: "Left out",
  confirmed: "Confirmed",
  declined: "Declined",
  dropped: "Dropped",
};

export const RSVP_LABEL: Record<RsvpStatus, string> = {
  going: "Said yes",
  not_going: "Said no",
  maybe: "Maybe",
  invited: "Not answered",
};

export const AVAILABILITY_LABEL: Record<AvailabilityStatus, string> = {
  available: "Free that day",
  maybe: "Might be free",
  unavailable: "Busy that day",
};

/// In the XI or on the bench — everyone the captain has committed to.
export function inSquad(state: SelectionState): boolean {
  return state === "selected" || state === "reserve" || state === "confirmed";
}

/// The order a captain reads the pool in: who said yes first, then who is
/// probably free, then everybody else — and within each, whoever has been left
/// out most often. Somebody available every week and never picked is the thing
/// a captain most needs pushed under their nose.
export function pickingOrder(a: Candidate, b: Candidate): number {
  const rank = (c: Candidate) =>
    c.rsvp === "going" ? 0
    : c.availability === "available" ? 1
    : c.rsvp === "maybe" || c.availability === "maybe" ? 2
    : c.rsvp === "not_going" || c.availability === "unavailable" ? 4
    : 3;
  const byRank = rank(a) - rank(b);
  if (byRank !== 0) return byRank;
  const byMissed = b.games_missed_out - a.games_missed_out;
  if (byMissed !== 0) return byMissed;
  return b.reliability_score - a.reliability_score;
}
